pkg/apis/audit/types.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package audit
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+)
+
+// Level defines the amount of information logged during auditing
+type Level string
+
+// Valid audit levels
+const (
+	// LevelNone disables auditing
+	LevelNone Level = "None"
+	// LevelMetadata provides the basic level of auditing.
+	LevelMetadata Level = "Metadata"
+	// LevelRequestObject provides Metadata level of auditing, and additionally
+	// logs the request object (does not apply for non-resource requests).
+	LevelRequestObject Level = "RequestObject"
+	// LevelResponseObject provides RequestObject level of auditing, and additionally
+	// logs the response object (does not apply for non-resource requests).
+	LevelResponseObject Level = "ResponseObject"
+)
+
+// Event captures all the information that can be included in an API audit log.
+type Event struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
+
+	// AuditLevel at which event was generated
+	Level Level
+
+	// Time the request reached the apiserver.
+	Timestamp metav1.Time
+	// Unique audit ID, generated for each request.
+	// +optional
+	RequestID string
+	// RequestURI is the request URI as sent by the client to a server.
+	RequestURI string
+	// Verb is the kubernetes verb associated with the request.
+	// For non-resource requests, this is identical to HttpMethod.
+	Verb string
+	// Authenticated user information.
+	User authentication.UserInfo
+	// Impersonated user information.
+	// +optional
+	Impersonate *authentication.UserInfo
+	// Source IP, from where the request originates.
+	// +optional
+	SourceIP string
+	// Object reference this request is targeted at.
+	// Does not apply for List-type requests, or non-resource requests.
+	// +optional
+	ObjectRef *api.ObjectReference
+	// The response status, populated even when the ResponseObject is not a Status type.
+	// For successful responses, this will only include the Code and StatusSuccess.
+	// For non-status type error responses, this will be auto-populated with the error Message.
+	// +optional
+	ResponseStatus *metav1.Status
+
+	// RequestObject logged before admission. Omitted for non-resource requests.
+	// Only logged at RequestObject Level and higher.
+	// +optional
+	RequestObject runtime.Object
```

Comment:
It's not serialized, it's only converted from internal to external. This means it's not json yet. I.e. the audit backend has to serialize them into whatever format it prefers.

In the proposal we have runtime.Unstructured. This would mean that the backend gets parsed essentially json. Though this needs two conversions to create (externalal -> json blob -> Unstructured).

Having runtime.Object here is cheap as we have to create that anyway. The audit backend can whatever it likes with it.